### PR TITLE
Convert excerpt to Preact (2/n)

### PR DESCRIPTION
**Pull request by robertknight**
**~~Depends on https://github.com/hypothesis/client/pull/1477~~**

Convert the `Excerpt` component to Preact. This is the control which wraps the annotation quote and body to limit their height, displaying expand/collapse controls if they exceed a threshold.

Rather than convert the existing implementation verbatim, this is a completely new implementation which should be easier to understand and use. Instead of requiring callers to provide an input property which represents the displayed data, which triggers a re-measurement of the content height if it changes, the new implementation observes the DOM directly for size changes.

This component renders caller-provided content (ie. it accepts a `children`
prop), which is not supported by the Preact <-> Angular bridge. Therefore it was
also necessary to create components (`AnnotationBody`, `AnnotationQuote`) that
encapsulate uses of `Excerpt` inside the `<annotation>` component.

 - #1477 adds a utility module to watch for changes in the size of a DOM node
 - Add new `Excerpt` implementation and remove the old one
 - Remove `excerpt-overflow-monitor` utility that is not used by the new
   implementation
 - Add `AnnotationBody` component to render an annotation's markup body
   inside a (new) excerpt and convert the Angular template for
   `<annotation>` (annotation.html) to use it.
 - Add `AnnotationQuote` component to render an annotation's quote
   inside an excerpt and convert `annotation.html` to use it

### CODE REVIEW

File: src/sidebar/components/excerpt.js [hunk 1]
```diff
@@ -1,178 +1,189 @@
...
-  this.contentStyle = overflowMonitor.contentStyle;
+  // Container for the excerpt's content.
+  const contentElement = useRef(null);
+
+  // Measured height of `contentElement` in pixels.
+  const [contentHeight, setContentHeight] = useState(0);
+
+  // Update the measured height of the content after the initial render and
+  // when the size of the content element changes.
+  const updateContentHeight = useCallback(() => {
```
Comment (lyzadanger) [inline src/sidebar/components/excerpt.js]: This may just be because the logic is what it is (and as simplified as it already can be), but it took me a while to grok the whole picture between here (line 80) and about line 96—the flow of state, control and what triggers what.

Part of this is simply that the `useCallback` hook is somewhat weirdly-named, and I always have to remind myself that's it's basically just memo-izing a fn reference.

I'm not sure I have any _suggestions_ here, it may be as straightforward as it can be already!

File: src/sidebar/components/excerpt.js [hunk 2]
```diff
@@ -1,178 +1,189 @@
...
+    const newContentHeight = contentElement.current.clientHeight;
+    setContentHeight(newContentHeight);
+
+    const isCollapsible =
+      newContentHeight > collapsedHeight + overflowHysteresis;
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/excerpt.js]: It turns out prettier removes them again if you do this, presumably because the parens do not appear in the AST. I agree about readability in this context, so I made use of [`// prettier-ignore`](https://prettier.io/docs/en/ignore.html).

File: src/sidebar/components/excerpt.js [hunk 3]
```diff
@@ -1,178 +1,189 @@
...
+    onCollapsibleChanged({ collapsible: isCollapsible });
+  }, [collapsedHeight, onCollapsibleChanged, overflowHysteresis]);
+
+  useLayoutEffect(() => {
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/excerpt.js]: `useLayoutEffect` effects run before the element is painted, `useEffect` effects run afterwards. In this case we need to calculate the collapsed state before the user sees the component as otherwise they will see a flash of the component in the wrong state.

In general, any situation where a component needs to examine its DOM output (eg. to measure its size), and then possibly change the output as a result, should use `useLayoutEffect`.

See https://reactjs.org/docs/hooks-reference.html#timing-of-effects.

File: src/sidebar/components/excerpt.js [hunk 4]
```diff
@@ -1,178 +1,189 @@
...
+      contentElement.current,
+      updateContentHeight
+    );
+    updateContentHeight();
+    return cleanup;
+  }, [updateContentHeight]);
+
+  // Render the (possibly truncated) content and controls for
+  // expanding/collapsing the content.
+  const overflowing = contentHeight > collapsedHeight + overflowHysteresis;
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/excerpt.js]: No, just an inconsistency. I think my general practice is to add an `is` or `has` prefix if it is not "sufficiently obvious" from the rest of the name that it is a boolean. I'll add an `is` prefix for consistency.

File: src/sidebar/components/excerpt.js [hunk 5]
```diff
@@ -1,178 +1,189 @@
...
-    overflowMonitor.check,
-    false /* capture */
+  const setCollapsed = collapsed =>
+    inlineControls
+      ? setCollapsedByInlineControls(collapsed)
+      : onToggleCollapsed(collapsed);
+
+  return (
+    <div className="excerpt" style={contentStyle}>
+      <div test-name="excerpt-content" ref={contentElement}>
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/excerpt.js]: It has no special meaning. I just added it as a convenience to be able to find this element in tests, since the `<div>` has no other distinguishing attributes.

File: src/sidebar/components/excerpt.js [hunk 6]
```diff
@@ -1,178 +1,189 @@
...
+ * The collapsed state of the container can be handled either via internal
+ * controls (if `inlineControls` is `true`) or by the caller using the
+ * `collapse` prop.
+ */
+function Excerpt({
+  children,
+  collapse = false,
+  collapsedHeight,
+  inlineControls = true,
+  onCollapsibleChanged = noop,
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/excerpt.js]: I think it depends on how the callback is used. If we only guard a call to it (eg. `if (fn) { fn(...) }`) this is quite useful. If on the other hand it triggers other logic, I wouldn't use it.

File: src/sidebar/components/annotation.js [hunk 7]
```diff
@@ -562,10 +562,6 @@ function AnnotationController(
       return;
     }
     self.canCollapseBody = canCollapse;
-
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation.js]: The Angular <-> Preact wrapper ensures that any callbacks (ie. where a Preact component invokes a callback prop passed in by an Angular component) trigger a digest. See https://github.com/hypothesis/client/blob/b43ed1abf20b1b815a15b22124c084df8a9cb579/src/sidebar/util/wrap-react-component.js#L36

File: src/sidebar/components/excerpt.js [hunk 8]
```diff
@@ -1,178 +1,189 @@
...
+/**
+ * A container which truncates its content when they exceed a specified height.
+ *
+ * The collapsed state of the container can be handled either via internal
+ * controls (if `inlineControls` is `true`) or by the caller using the
+ * `collapse` prop.
+ */
+function Excerpt({
+  children,
+  collapse = false,
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/excerpt.js]: I expanded the `collapse` prop docs to clarify that it is used to control the collapsed/expanded state if the Excerpt is configured not to provide internal controls for this.

**Comment by robertknight** [comment]
Hi @lyzadanger. Thanks for the initial review. I have updated the PR accordingly.

> For my personal ergonomic comfort (if I were going to be the reviewer), one place to start might be to calve off just the observe-element-size util and its tests into a PR as that would give me a chance to dive into the nuts and bolts of that (and make sure my ResizeObserver and MutationObserver fu is up to snuff). But maybe the whole shebang is fine.

I'm happy to carve off that part if you'd prefer. Just give me a ping on Slack and let me know your preference.

**Comment by lyzadanger** [comment]
@robertknight If you calve off that `observer` PR, I'll get it a speedy review.

**Comment by robertknight** [comment]
Rebased. I could potentially split out the `Excerpt` implementation into a separate PR. Let me know if you want me to do that.

File: src/sidebar/components/excerpt.js [hunk 9]
```diff
@@ -1,178 +1,193 @@
...
+function InlineControls({ isCollapsed, setCollapsed, linkStyle = {} }) {
+  const toggleTitle = isCollapsed
+    ? 'Show the full excerpt'
+    : 'Show the first few lines only';
+  const toggleLabel = isCollapsed ? 'More' : 'Less';
+
+  return (
+    <div className="excerpt__inline-controls">
+      <span className="excerpt__toggle-link">
+        <a
```
Comment (lyzadanger) [inline src/sidebar/components/excerpt.js]: Wondering if a `<button>` element would make more sense here?

File: src/sidebar/components/excerpt.js [hunk 10]
```diff
@@ -1,178 +1,193 @@
...
-     * the excerpt if it has only been truncated by a very small amount, such
-     * that expanding the excerpt would reveal no extra lines of text.
-     */
-    overflowHysteresis: '<?',
-  },
-  transclude: true,
-  template: require('../templates/excerpt.html'),
+Excerpt.propTypes = {
+  /**
+   * The content to render inside the container.
```
Comment (lyzadanger) [inline src/sidebar/components/excerpt.js]: I feel like we could probably get away with not commenting `children` props if we didn't feel like it. That said, I have no aversion to this comment.

File: src/sidebar/components/test/annotation-body-test.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+  beforeEach(() => {
+    AnnotationBody.$imports.$mock(mockImportedComponents());
+  });
+
+  afterEach(() => {
+    AnnotationBody.$imports.$restore();
+  });
+
+  it('displays the comment if `isEditing` is false', () => {
```
Comment (lyzadanger) [inline src/sidebar/components/test/annotation-body-test.js]: "the comment" meaning "the annotation body content?"

File: src/sidebar/components/test/excerpt-test.js [hunk 12]
```diff
@@ -1,209 +1,135 @@
...
+      <Excerpt
+        collapse={true}
+        collapsedHeight={40}
+        inlineControls={false}
+        settings={{}}
+        {...props}
+      >
+        {content}
+      </Excerpt>,
+      { attachTo: container }
```
Comment (lyzadanger) [inline src/sidebar/components/test/excerpt-test.js]: Ooh, this `attachTo` `option` is handy.

File: src/sidebar/index.js [hunk 13]
```diff
@@ -160,14 +167,6 @@ function startAngularApp(config) {
       'loggedOutMessage',
       wrapReactComponent(require('./components/logged-out-message'))
     )
-    .component(
```
Comment (lyzadanger) [inline src/sidebar/index.js]: Good job remembering to remove these.

File: src/sidebar/components/annotation-quote.js [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+AnnotationQuote.propTypes = {
+  /**
+   * If `true`, display an indicator that the annotated text was not found in
+   * the current version of the document.
+   */
+  isOrphan: propTypes.bool,
+
+  /**
+   * The text that the annotation refers to. This is rendered as plain text
+   * (ie. HTML tags are rendered literally).
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation-quote.js]: Yes, exactly.

File: src/sidebar/components/annotation-body.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+AnnotationBody.propTypes = {
+  /**
+   * Whether to limit the height of the annotation body if it is tall.
+   */
+  collapse: propTypes.bool,
+
+  /**
+   * If the
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation-body.js]: Looking at this again, the `hasContent` prop is badly named and possibly redundant. It controls whether the annotation body displays struck-through text or a "redacted text" placeholder (a series of grey bars) if the annotation has been moderated. I will add a comment for the moment and then address in a follow-up.

File: src/sidebar/components/annotation-body.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        </Excerpt>
+      )}
+      {isEditing && <MarkdownEditor text={text} onEditText={onEditText} />}
+    </section>
+  );
+}
+
+AnnotationBody.propTypes = {
+  /**
+   * Whether to limit the height of the annotation body if it is tall.
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation-body.js]: Yes, I'll update the doc.

File: src/sidebar/components/annotation-body.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+  /**
+   * `true` if the contents of this annotation body have been redacted by
+   * a moderator.
+   */
+  isHiddenByModerator: propTypes.bool,
+
+  /**
+   * Callback invoked when the height of the rendered annotation body increases
+   * above or falls below the threshold at which the `collapse` prop will affect
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation-body.js]: I thought it useful to explicitly mention the `collapse` prop here in case someone wonders "what threshold"?

File: src/sidebar/components/annotation-body.js [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+   * it.
+   */
+  onCollapsibleChanged: propTypes.func,
+
+  /**
+   * Callback invoked when the user edits the content of the annotation body
+   * when `isEditing` is true.
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation-body.js]: Indeed. I'll remove the part of the sentence.

File: src/sidebar/components/annotation-body.js [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+   */
+  onEditText: propTypes.func,
+
+  /**
+   * Callback invoked when the user clicks a space in a truncated annotation
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation-body.js]: I'll reword this to clarify.

File: src/sidebar/components/annotation-body.js [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+  return (
+    <section className="annotation-body">
+      {!isEditing && (
+        <Excerpt
+          collapse={collapse}
+          collapsedHeight={400}
+          inlineControls={false}
+          onCollapsibleChanged={onCollapsibleChanged}
+          onToggleCollapsed={collapsed => onToggleCollapsed({ collapsed })}
+          overflowHystersis={20}
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation-body.js]: Gah. The term "hysteresis" was originally chosen as in "thermostat hysteresis". However, per Wikipedia, "Hysteresis is the dependence of the state of a system on its history". This prop doesn't really work like that though. It is just a fixed threshold above `collapsedHeight` before the excerpt is collapsed. I've renamed it to `overflowThreshold` instead.

In practice, it allows a ~4 line quote to be shown fully even though quotes are collapsed to 3 lines if they exceed that height. For annotation bodies, it is less important as the collapsed height is much taller.

File: src/sidebar/components/annotation-quote.js [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+const { applyTheme } = require('../util/theme');
+const Excerpt = require('./excerpt');
+
+/**
+ * Display the selected text from the document associated with an annotation.
+ */
+function AnnotationQuote({ isOrphan, quote, settings = {} }) {
+  return (
+    <section
+      className={classnames('annotation-quote-list', isOrphan && 'is-orphan')}
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation-quote.js]: Looking at the API, an annotation can have multiple targets or multiple quote selectors within a target. AFAIK in Firefox it is, or was, technically possible to create a discontinuous text selection. We don't really support this though so it is obsolete.

File: src/sidebar/components/annotation-body.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/annotation-body.js]: The value was chosen to roughly correspond to the height of a line in an annotation quote / body. If you were using it in a context with a larger font or a different type of content you might want a different value. IIRC the reason it isn't measured in units of line height is purely for reasons of implementation ease.

File: src/sidebar/components/excerpt.js (hunk 10 above)
Comment (robertknight) [reply src/sidebar/components/excerpt.js]: I agree. I'll remove it.

File: src/sidebar/components/test/annotation-body-test.js (hunk 11 above)
Comment (robertknight) [reply src/sidebar/components/test/annotation-body-test.js]: Yes.

**Comment by robertknight** [comment]
> Feel free to merge once the few actual action items (incomplete comments, Hysteresis typos) are addressed to your satisfaction.

Thanks Lyza. I've addressed the comments and will merge when tests complete. I made a note of some additional possible simplifications which I'll do as a follow-up.
